### ISSUE TITLE
Router Replay (offline evaluation + data flywheel)

### ISSUE THREAD

**Issue by Xunzhuo** [vllm-project/semantic-router#539]
### Summary

Introduce **Router Replay** to vLLM Semantic Router (vLLM-SR): a structured “decision log + replayer” that records each routing decision (state → candidates → policy → action → outcome → eval) and allows **offline replay** to evaluate new policies and to produce training data for imitation, preference learning, and offline RL—without touching live traffic.

---

### Motivation

* **Safe iteration:** Evaluate new routing strategies offline before any canary traffic.
* **Better policies over time:** Convert production decisions into a **data flywheel** for behavior cloning, DPO/IPO, and conservative offline RL (IQL/CQL).
* **Explainability & audit:** Reconstruct “why we routed this way” with traceable evidence.
* **Multi-objective control:** Optimize quality ⇄ cost ⇄ latency ⇄ safety using a unified reward.

---

### Goals

1. Log a minimally sufficient, privacy-aware record per routing decision.
2. Provide a **Replayer** to run historical states through alternative policies and compute counterfactual metrics (IPS/SNIPS/DR).
3. Ship **Evaluators** (quality/cost/latency/safety) for offline scoring.
4. Expose **training hooks** to export replay data to SL/DPO/Offline-RL pipelines.
5. Make it observable (OpenTelemetry spans) and production-friendly (batching, backpressure, retention, TTL).

### Non-Goals

* Not a response cache.
* Not online RL or full A/B platform (this lays the groundwork).

---

#### 1) Data Model (JSON event; Parquet recommended for storage)

```json
{
  "ts": "2025-10-26T12:00:00Z",
  "session_id": "s-123",
  "turn_id": 3,

  "state": {
    "intent": "math.calculus",
    "tags": ["pii_safe","jailbreak_ok"],
    "features": {"len_in": 312, "lang": "en", "complexity_est": 0.62},
    "embedding_id": "vec_abc123"              // pointer; do not store raw text
  },

  "candidates": [
    {"id": "qwen-14b-math-pro", "params": {"temp": 0.2}, "price": 2.0, "latency_p50": 600},
    {"id": "gpt-x-general", "params": {"temp": 0.1}}
  ],

  "policy": {
    "version": "sr-0.1.7",
    "probs": {"qwen-14b-math-pro": 0.62, "gpt-x-general": 0.38}
  },

  "action": {"route": "qwen-14b-math-pro", "params": {"temp": 0.2}},

  "outcome": {
    "tokens_in": 312, "tokens_out": 210,
    "latency_ms": 812, "cost_usd": 0.0007,
    "guardrail_ok": true
  },

  "eval": {
    "task_score": 0.90,            // from automatic evaluator or test harness
    "preference": "win_vs_baseline",
    "reward": 0.83                 // normalized composite: quality - cost - latency - violations
  },

  "privacy": {"redacted": true, "pii_policy": "drop_text_keep_stats"}
}
```

Recommended Parquet columns mirror the above (flat or nested), partitioned by date and intent.

#### 2) Logging Interface

* **DecisionLogger** (async, loss-tolerant, back-pressured; batch to Kafka or local queue → object store).
* Configurable **redaction**: drop raw text by default; store feature stats, hashed keys, or vector IDs.
* OpenTelemetry span linking: `router.decision` with attributes for `state.intent`, `action.route`, `latency_ms`, etc.

#### 3) Storage & Retention

* Write-ahead queue (Kafka/NATS/FS buffer) → **Parquet/Delta** in object storage.
* Retention + TTL + per-tenant encryption and deletion hooks (“right to be forgotten”).

#### 4) Replayer

* CLI/SDK: feed a policy implementation (or JSON policy) and run on historical `state + candidates`.
* Supports **counterfactual evaluation**:

  * IPS/SNIPS (inverse propensity weighting),
  * Doubly-Robust (DR) with learned outcome models (for latency/cost/quality).
* Emits offline report: uplift curves, risk bands, intent-level breakdown.

#### 5) Evaluators

* **Quality**: task-specific autoscorers, regression tests, or LLM-judge with consistency checks.
* **Cost/Latency**: measured if available; otherwise model-based estimators calibrated from logs.
* **Safety**: guardrail signals → hard penalties.

#### 6) Training Hooks

* **SL/Behavior Cloning**: learn from “good” actions.
* **Preference Learning**: DPO/IPO from winner/loser pairs.
* **Offline RL**: IQL/CQL/CRR with conservative penalties to avoid out-of-distribution collapse.
* Export datasets as Arrow/Parquet; provide dataloaders.

**Comment by Xunzhuo** [comment]
this is done
